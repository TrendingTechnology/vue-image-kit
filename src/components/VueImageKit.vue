<template>
  <div class="vue-image-kit">
    <div v-if="dataUrl" class="vue-image-kit__placeholder" :style="{ backgroundColor }">
      <img :src="placeholder || dataUrl" alt="Placeholder" :style="{ width: `${width}px`, height: `${height}px` }"/>
    </div>
    <img class="vue-image-kit__img" :srcset="getSrcset" :sizes="getSizes" :src="getSrc" :alt="alt" :style="{ width: `${width}px`, height: `${height}px` }"/>
  </div>
</template>

<script>
export default {
  name: 'VueImageKit',
  props: {
    hash: {
      type: String,
      required: true
    },
    src: {
      type: String,
      required: true
    },
    placeholder: {
      type: String,
      default: ''
    },
    backgroundColor: {
      type: String,
      default: ''
    },
    width: {
      type: Number,
      default: null
    },
    height: {
      type: Number,
      default: null
    },
    alt: {
      type: String,
      default: ''
    },
    srcset: {
      type: Array,
      default: () => [320, 480, 800]
    },
    sizes: {
      type: Array,
      default: () => []
    },
    defaultSize: {
      type: Number,
      default: 1080
    },
    customTransform: {
      type: String,
      default: ''
    }
  },
  data: () => ({
    imageKitPrefix: 'https://ik.imagekit.io',
    showCanvas: true,
    observer: null,
    timeOut: null
  }),
  computed: {
    dataUrl () {
      const { width, height, showCanvas } = this

      if (!width || !height || !showCanvas) return ''

      const w = 100
      const canvas = document.createElement('canvas')
      canvas.width = w
      canvas.height = (height / width) * w

      return canvas.toDataURL()
    },
    getSrc () {
      const { showCanvas, dataUrl, imageKitPrefix, hash, src } = this

      return showCanvas
        ? dataUrl
        : `${imageKitPrefix}/${hash}/${src}`
    },
    getSrcset () {
      const { srcset, imageKitPrefix, hash, customTransform, src } = this

      return srcset
        .map(size => `${imageKitPrefix}/${hash}/tr:w-${size}${customTransform ? ',' + customTransform : ''}/${src} ${size}w`)
        .join(', ')
    },
    getSizes () {
      const { sizes, srcset, defaultSize } = this
      let sizesString = ''
      if (sizes && sizes.length && (sizes.length === srcset.length)) {
        sizesString = srcset.map((size, idx) => `(max-width: ${size}px) ${sizes[idx]}px`).join(', ')
      }

      sizesString = srcset.map(size => `(max-width: ${size}px) ${parseInt(size, 10) - 40}px`).join(', ')
      sizesString += ` ${defaultSize}px`

      return sizesString
    }
  },
  mounted () {
    this.showCanvas = true
    this.observer = new IntersectionObserver(([entry]) => {
      this.triggerIntersection(entry)
    })
    this.observer.observe(this.$el)
    this.$once('hook:beforeDestroy', () => {
      this.observer.disconnect()

      if (this.timeOut) {
        clearTimeout(this.timeOut)
      }
    })
  },
  methods: {
    triggerIntersection(entry = {}) {
      const { $el, timeOut, srcset, getSrcset, imageKitPrefix, hash, src } = this
      const img = $el.querySelector('.vue-image-kit__img')
      const placeholder = $el.querySelector('.vue-image-kit__placeholder')

      img.onload = function () {
        delete img.onload
        $el.classList.add('vue-image-kit--loaded')

        if (placeholder) {
          timeOut = setTimeout(() => {
            placeholder.remove()
          }, 300)
        }
      }

      if (entry.isIntersecting) {
        this.showCanvas = false

        if (srcset) {
          img.srcset = getSrcset
        }

        img.src = `${imageKitPrefix}/${hash}/${src}`
        this.observer.disconnect()
      }
    }
  }
}
</script>

<style scoped>
.vue-image-kit {
  display: inline-block;
  position: relative;
}

.vue-image-kit__placeholder {
  position: absolute;
  overflow: hidden;
}

.vue-image-kit__placeholder img {
  transform: scale(1.05);
  filter: blur(10px);
}

.vue-image-kit__img {
  opacity: 0;
  transition: opacity 300ms ease;
}

.vue-image-kit--loaded .vue-image-kit__img {
  opacity: 1;
}
</style>
