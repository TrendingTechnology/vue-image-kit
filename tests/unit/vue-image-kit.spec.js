import { mount } from '@vue/test-utils'
import VueImageKit from '../../src/components/VueImageKit'
import 'jest-canvas-mock'

function setupIntersectionObserverMock({
  observe = () => null,
  unobserve = () => null,
} = {}) {
  class IntersectionObserver {
    observe = observe;
    unobserve = unobserve;
  }
  Object.defineProperty(
    window,
    'IntersectionObserver',
    { writable: true, configurable: true, value: IntersectionObserver }
  );
  Object.defineProperty(
    global,
    'IntersectionObserver',
    { writable: true, configurable: true, value: IntersectionObserver }
  );
}

describe('When I create the VueImageKit component', () => {
  const item = { hash: '6xhf1gnexgdgk', src: 'lion_BllLvaqVn.jpg' }

  beforeEach(() => {
    setupIntersectionObserverMock();
  })
  const createComponent = (propsData = {}, slot) => {
    return mount(VueImageKit, {
      propsData
    })
  }

  it('should be a Vue instance', () => {
    const wrapper = createComponent(item)
    expect(wrapper.find(VueImageKit).isVueInstance()).toBe(true)
  })

  it('should validate all props', () => {
    const consoleLog = console.error
    console.error = jest.fn()
    const wrapper = createComponent(item)
    const hash = wrapper.vm.$options.props.hash
    expect(hash.required).toBeTruthy()
    expect(hash.type).toBe(String)
    expect(hash.default).toBe(undefined)
    const src = wrapper.vm.$options.props.src
    expect(src.required).toBeTruthy()
    expect(src.type).toBe(String)
    expect(src.default).toBe(undefined)
    const placeholder = wrapper.vm.$options.props.placeholder
    expect(placeholder.required).toBeFalsy()
    expect(placeholder.type).toBe(String)
    expect(placeholder.default).toBe('')
    const backgroundColor = wrapper.vm.$options.props.backgroundColor
    expect(backgroundColor.required).toBeFalsy()
    expect(backgroundColor.type).toBe(String)
    expect(backgroundColor.default).toBe('')
    const srcset = wrapper.vm.$options.props.srcset
    expect(srcset.required).toBeFalsy()
    expect(srcset.type).toBe(Array)
    const defaultSrcset = { default: () => [320, 480, 800] }
    expect(JSON.stringify(srcset.default)).toEqual(JSON.stringify(defaultSrcset.default))
    const sizes = wrapper.vm.$options.props.sizes
    expect(sizes.required).toBeFalsy()
    expect(sizes.type).toBe(Array)
    console.log(JSON.stringify(sizes.default))
    expect(JSON.stringify(sizes.default)).toBeFalsy()
    const defaultSize = wrapper.vm.$options.props.defaultSize
    expect(defaultSize.required).toBeFalsy()
    expect(defaultSize.type).toBe(Number)
    expect(defaultSize.default).toBe(1080)
    const customTransform = wrapper.vm.$options.props.customTransform
    expect(customTransform.required).toBeFalsy()
    expect(customTransform.type).toBe(String)
    expect(customTransform.default).toBe('')
    const width = wrapper.vm.$options.props.width
    expect(width.required).toBeFalsy()
    expect(width.type).toBe(Number)
    expect(width.default).toBe(null)
    const height = wrapper.vm.$options.props.height
    expect(height.required).toBeFalsy()
    expect(height.type).toBe(Number)
    expect(height.default).toBe(null)
    const alt = wrapper.vm.$options.props.alt
    expect(alt.required).toBeFalsy()
    expect(alt.type).toBe(String)
    expect(alt.default).toBe('')
    const component = wrapper.find('.vue-image-kit')
    expect(component.exists()).toBe(true)
    console.error = consoleLog
  })

  it('should have a width, height and a placeholder', () => {
    const wrapper = createComponent({ ...item, width: 300, height: 300, placeholder: 'https://ik.imagekit.io/6xhf1gnexgdgk/igor2_HJhiHMa54.png' })
    const found = wrapper.find('.vue-image-kit > .vue-image-kit__placeholder > img')
    expect(found.exists()).toBe(true)
    const expected = {
      src: 'https://ik.imagekit.io/6xhf1gnexgdgk/igor2_HJhiHMa54.png',
      alt: 'Placeholder',
      style: 'width: 300px; height: 300px;'
    }
    expect(JSON.stringify(found.attributes())).toBe(JSON.stringify(expected))
    const main = wrapper.find('.vue-image-kit > .vue-image-kit__img')
    expect(main.exists()).toBe(true)
    expect(main.attributes().style).toBe('width: 300px; height: 300px;')
  })

  it('should have a alt attribute', () => {
    const wrapper = createComponent({ ...item, alt: 'Lion image' })
    const main = wrapper.find('.vue-image-kit > .vue-image-kit__img')
    expect(main.exists()).toBe(true)
    expect(main.attributes().alt).toBe('Lion image')
  })

  it('should have background color in the placeholder', () => {
    const wrapper = createComponent({ ...item, width: 300, height: 300, placeholder: 'https://ik.imagekit.io/6xhf1gnexgdgk/igor2_HJhiHMa54.png', backgroundColor: '#f40' })
    const placeholder = wrapper.find('.vue-image-kit > .vue-image-kit__placeholder')
    expect(placeholder.exists()).toBe(true)
    expect(placeholder.attributes().style).toBe('background-color: rgb(255, 68, 0);') // #f40
  })

  // ! Missing tests for props
  // srcset: {
  //   type: Array,
  //   default: () => [320, 480, 800]
  // },
  // sizes: {
  //   type: Array,
  //   default: () => []
  // },
  // defaultSize: {
  //   type: Number,
  //   default: 1080
  // },
  // customTransform: {
  //   type: String,
  //   default: ''
  // }

  it('should match snapshot', () => {
    const wrapper = createComponent({
      ...item,
      placeholder: '',
      backgroundColor: '',
      width: null,
      height: null,
      alt: '',
      srcset: [320, 480, 800],
      sizes: [],
      defaultSize: 1080,
      customTransform: ''
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
