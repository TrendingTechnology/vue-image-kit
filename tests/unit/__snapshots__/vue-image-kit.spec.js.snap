// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the VueImageKit component should match snapshot 1`] = `
<div class="vue-image-kit">
  <!----> <img sizes="(max-width: 320px) 280px, (max-width: 480px) 440px, (max-width: 800px) 760px 1080px" alt="" class="vue-image-kit__img"></div>
`;
